gui/default/syncthing/folder/ignoresService.js

```diff
@@ -0,0 +1,257 @@
+        // by prefix (for simple patterns)
+        // The changes made to the path make it more specific than the original
+        // pattern. It is useful for prefix matcher. Should a glob matcher be
+        // added this path should not be used to match because it does not
+        // represent the complete pattern. The most common effect of the changes
+        // is matching patterns like `*`, `**`, or `/*` without a special case
+        function toPath(line) {
+            // Add a leading slash when the pattern begins with wildcards
+            // because a pattern beginning with * or ** will match entries at
+            // the root (as well as those in child directories)
+            line = line.replace(/^\*+/, '/$&'); // `$&` inserts the matched substring
+
+            // Trim trailing wildcards after separator because child paths would
+            // already be prefixed by this path
+            line = line.replace(/\/\*+$/, '/');
+            return line;
+        }
+
+        // Inspect the pattern to determine if it is simple enough to match,
+        // then return a function that can be applied to match a path.
+        function chooseMatcher(line) {
+            if (line.length === 0) return neverMatch;
+            if (line.indexOf('//') === 0) return neverMatch; // comment
+            if (line.indexOf('/') !== 0) return null; // not a root line
+            if (line.length > 1 && line.charAt(line.length - 1) === '/') return null; // trailing slash
+
+            line = line.replaceAll(/\\[\*\?\[\]\{\}]/g, '') // remove properly escaped glob characters
+            if (line.match(/[\*\?\[\]\{\}]/)) {
+                // The pattern contains a special character. The pattern is a
+                // glob, too complex for us to handle in the UI. Return no
+                // matcher function indicating we cannot match it.
+                return null;
+            }
+
+            return prefixMatch;
+        }
+
+        function findLastIndex(array, predicate) {
+            var i = array.length;
+            while (i--) {
+                if (predicate(array[i], i, array)) {
+                    return i;
+                }
+            }
+            return -1;
+        }
+
+        // Here we abuse matchFunc by using it to match another
+        // pattern's path. The comparison assumes both patterns are
+        // simple with a prefixMatch type, so if support for glob
+        // patterns is added this will need a different approach.
+        function patternMoreSpecificThan(patternA, patternB) {
+            return patternB.isSimple && patternA.isSimple && patternB.matchFunc(patternA.path)
+        }
+
+        /*
+         * pattern matching functions
+         */
+
+        // A prefix match is anchored at folder root and is an exact match to
+        // one file or directory
+        function prefixMatch(filePath) {
+            var patternPath = this.path
```

Comment:
This is magical, I assume this is some prototype voodoo that adds the method on the object?

Reply:
Yes, `this` is magical :) Technically prototype isn't involved here, simply calling a function on an object sets `this` to the object it's called on. It can be bad form to use `this` when the context is ambiguous, but in this case the matchFunc assignment and call are done in a way that is [clearly defined](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/this#as_an_object_method) so I'm okay with it.
(An alternative approach might be to use `prefixMatch.call(pattern, filePath)` to set `this` to `pattern` explicitly, but that has the same effect as `pattern.prefixMatch(filePath)` while being harder to read)